### ISSUE TITLE
how to disable nsfw filter

### ISSUE THREAD

**Issue by loboere** [lkwq007/stablediffusion-infinity#8]
I want to draw things that are not nsfw but it detects me as nsfw and draws me a black square.

**Comment by diegoegus1** [comment]
same

**Comment by realdep** [comment]
yep, having the same issue.

**Comment by devilkkw** [comment]
go to
..\sd-inf\Lib\site-packages\diffusers\pipelines\stable_diffusion\
edit safety_checker.py (make a backup copy) and replace with:

`import numpy as np
import torch
import torch.nn as nn

from transformers import CLIPConfig, CLIPVisionModel, PreTrainedModel

from ...utils import logging


logger = logging.get_logger(__name__)


def cosine_distance(image_embeds, text_embeds):
    normalized_image_embeds = nn.functional.normalize(image_embeds)
    normalized_text_embeds = nn.functional.normalize(text_embeds)
    return torch.mm(normalized_image_embeds, normalized_text_embeds.t())


class StableDiffusionSafetyChecker(PreTrainedModel):
    config_class = CLIPConfig

    def __init__(self, config: CLIPConfig):
        super().__init__(config)

        self.vision_model = CLIPVisionModel(config.vision_config)
        self.visual_projection = nn.Linear(config.vision_config.hidden_size, config.projection_dim, bias=False)

        self.concept_embeds = nn.Parameter(torch.ones(17, config.projection_dim), requires_grad=False)
        self.special_care_embeds = nn.Parameter(torch.ones(3, config.projection_dim), requires_grad=False)

        self.register_buffer("concept_embeds_weights", torch.ones(17))
        self.register_buffer("special_care_embeds_weights", torch.ones(3))

    @torch.no_grad()
    def forward(self, clip_input, images):
        pooled_output = self.vision_model(clip_input)[1]  # pooled_output
        image_embeds = self.visual_projection(pooled_output)

        special_cos_dist = cosine_distance(image_embeds, self.special_care_embeds).cpu().numpy()
        cos_dist = cosine_distance(image_embeds, self.concept_embeds).cpu().numpy()

        result = []
        batch_size = image_embeds.shape[0]
        for i in range(batch_size):
            result_img = {"special_scores": {}, "special_care": [], "concept_scores": {}, "bad_concepts": []}

            # increase this value to create a stronger `nfsw` filter
            # at the cost of increasing the possibility of filtering benign images
            adjustment = 0.0

            for concet_idx in range(len(special_cos_dist[0])):
                concept_cos = special_cos_dist[i][concet_idx]
                concept_threshold = self.special_care_embeds_weights[concet_idx].item()
                result_img["special_scores"][concet_idx] = round(concept_cos - concept_threshold + adjustment, 3)
                if result_img["special_scores"][concet_idx] > 0:
                    result_img["special_care"].append({concet_idx, result_img["special_scores"][concet_idx]})
                    adjustment = 0.01

            for concet_idx in range(len(cos_dist[0])):
                concept_cos = cos_dist[i][concet_idx]
                concept_threshold = self.concept_embeds_weights[concet_idx].item()
                result_img["concept_scores"][concet_idx] = round(concept_cos - concept_threshold + adjustment, 3)
                if result_img["concept_scores"][concet_idx] > 0:
                    result_img["bad_concepts"].append(concet_idx)

            result.append(result_img)

        has_nsfw_concepts = [len(res["bad_concepts"]) < 0 for res in result]

        for idx, has_nsfw_concept in enumerate(has_nsfw_concepts):
            if has_nsfw_concept:
                images[idx] = np.zeros(images[idx].shape)  # black image

        if any(has_nsfw_concepts):
            logger.warning(
                "Potential NSFW content was detected in one or more images. A black image will be returned instead."
                " Try again with a different prompt and/or seed."
            )

        return images, has_nsfw_concepts

    @torch.inference_mode()
    def forward_onnx(self, clip_input: torch.FloatTensor, images: torch.FloatTensor):
        pooled_output = self.vision_model(clip_input)[1]  # pooled_output
        image_embeds = self.visual_projection(pooled_output)

        special_cos_dist = cosine_distance(image_embeds, self.special_care_embeds)
        cos_dist = cosine_distance(image_embeds, self.concept_embeds)

        # increase this value to create a stronger `nsfw` filter
        # at the cost of increasing the possibility of filtering benign images
        adjustment = 0.0

        special_scores = special_cos_dist - self.special_care_embeds_weights + adjustment
        # special_scores = special_scores.round(decimals=3)
        special_care = torch.any(special_scores > 0, dim=1)
        special_adjustment = special_care * 0.01
        special_adjustment = special_adjustment.unsqueeze(1).expand(-1, cos_dist.shape[1])

        concept_scores = (cos_dist - self.concept_embeds_weights) + special_adjustment
        # concept_scores = concept_scores.round(decimals=3)
        has_nsfw_concepts = torch.any(concept_scores > 0, dim=1)

        images[has_nsfw_concepts] < 0.0  # black image

        return images, has_nsfw_concepts`

**Comment by kirilledelman** [comment]
> 

Do you mind giving an absolute path (on a Colab instance) of this file, because I can't find it.

**Comment by darthdeus** [comment]
There's a much simpler solution, you can just do this

```
text2img.safety_checker = lambda images, **kwargs: (images, False)
inpaint.safety_checker = lambda images, **kwargs: (images, False)
```

More info here https://write.farook.org/adventures-in-imagineering-stable-diffusion-the-unstable-bits/

**Comment by devilkkw** [comment]
> > 
> 
> Do you mind giving an absolute path (on a Colab instance) of this file, because I can't find it.

Sorry, there's not specified in opening post. for collab i don't know. i use it locally, the folder are located in your conda data folder, in my case is "C:\ProgramData\Miniconda3\envs\sd-inf\Lib\site-packages\diffusers\pipelines\stable_diffusion".
 for collab im sorry, i don't know if is possible to doing the same modification.

**Comment by bdszoke** [comment]
> > > 
> > 
> > 
> > Do you mind giving an absolute path (on a Colab instance) of this file, because I can't find it.
> 
> Sorry, there's not specified in opening post. for collab i don't know. i use it locally, the folder are located in your conda data folder, in my case is "C:\ProgramData\Miniconda3\envs\sd-inf\Lib\site-packages\diffusers\pipelines\stable_diffusion". for collab im sorry, i don't know if is possible to doing the same modification.

I wonder if the right way to go would be to create a separately hosted instance from which to access the files, either by enabling it to access a local drive, or by having it hosted online such as a separate repository or even fork. Unfortunately, I don't know enough about how Colab does things to know either :(

**Comment by darthdeus** [comment]
> > > > 
> > > 
> > > 
> > > Do you mind giving an absolute path (on a Colab instance) of this file, because I can't find it.
> > 
> > 
> > Sorry, there's not specified in opening post. for collab i don't know. i use it locally, the folder are located in your conda data folder, in my case is "C:\ProgramData\Miniconda3\envs\sd-inf\Lib\site-packages\diffusers\pipelines\stable_diffusion". for collab im sorry, i don't know if is possible to doing the same modification.
> 
> I wonder if the right way to go would be to create a separately hosted instance from which to access the files, either by enabling it to access a local drive, or by having it hosted online such as a separate repository or even fork. Unfortunately, I don't know enough about how Colab does things to know either :(

You. can just use the solution I posted, that is

```
text2img.safety_checker = lambda images, **kwargs: (images, False)
inpaint.safety_checker = lambda images, **kwargs: (images, False)
```

this owrks and requires no source code modification

**Comment by lkwq007** [comment]
As the updated UI has a checkbox for that, I'll close this issue.
